types/underscore/index.d.ts

```diff
@@ -106,22 +107,24 @@ declare module _ {
         (prev: TResult, curr: T, key: string, list: Dictionary<T>): TResult;
     }
 
+    type TypeOfList<V> = V extends _.List<infer T> ? T : never;
+
+    type TypeOfDictionary<T> = T extends _.Dictionary<infer V> ? V : never;
+
     interface Cancelable {
         cancel(): void;
     }
 
-    type TypeOfDictionary<T> = T extends _.Dictionary<infer V> ? V : never;
-
     interface UnderscoreStatic {
         /**
-        * Underscore OOP Wrapper, all Underscore functions that take an object
-        * as the first parameter can be invoked through this function.
-        * @param key First argument to Underscore object functions.
-        **/
-        <T>(value: _.List<T>): Underscore<T, _.List<T>>;
-        <T>(value: Array<T>): Underscore<T, Array<T>>;
-        <T extends TypeOfDictionary<V>, V extends _.Dictionary<any> = _.Dictionary<T>>(value: V): Underscore<T, V>;
-        <T>(value: T): Underscore<T>;
+         * Underscore OOP Wrapper, all Underscore functions that take an object
+         * as the first parameter can be invoked through this function.
+         * @param value First argument to Underscore object functions.
+         * @returns An Underscore wrapper around the supplied value.
+         **/
+        <T extends TypeOfList<V>, V extends List<any> = List<T>>(value: V): Underscore<T, V>;
```

Comment:
With regards to the performance impact of these changes, this will technically handle someone's custom extended array type, but that's probably not a common scenario. If this performance change is really a concern, I could change the Underscore and chain functions to use the below more static set of overloads instead and only be making this a tiny bit more inconvenient for those hopefully rare few that make lists or dictionaries that have extra bits tacked on.

```TS
        <T>(value: T[]): Underscore<T, T[]>;
        <T>(value: List<T>): Underscore<T, List<T>>;
        <T>(value: Dictionary<T>): Underscore<T, Dictionary<T>>;
        <V>(value: V): Underscore<V>;
```

Reply:
Still probably not a big deal, but I suppose one common item that fits the bill of a list that might have extra bits tacked on is `arguments` (though the only extra bit I see right now in Chrome is `callee`).

Reply:
There is one place that is immediately improved by these changes and two that would be improved in a future PR. The spot that is immediately improved is `.value()`, which will now provide the original wrapped type instead of an Underscore interface that may not exactly match the provided type. The first improvement in a later PR is `each`, which is expected to return the same object it was given. The second improvement in a later PR is the type of the original object that is provided as the third argument to iterators, though that will be applied to `Underscore` and `_Chain` but not `UnderscoreStatic` in the currently planned set of changes.

All that being said, it's not super useful to use the return value of `each` or to do something like `_(specialArray).value()` or `_.chain(specialArray).value()`, and it's probably not even terribly interesting to retain the original type for iterators (personally I've never actually written an iterator that's even used that third argument).

If the possibility of being given types that are assignable to but are not exactly the same as Underscore interfaces doesn't need to be handled, I don't think it's necessary to use the extra constraints for dictionaries either; `<T>(value: Dictionary<T>): Underscore<T, Dictionary<T>>;` should be a sufficient overload for dictionaries as long as it comes before the overload that takes any value. It seems like it would be nice to consistently treat special cases for both dictionaries and arrays in a similar fashion, but I will admit that special cases for dictionaries seem more likely to actually happen in the real world so I'd be ok with leaving `V extends Dictionary<T>` but removing the similar constraint for lists if you'd prefer.

Short contrived examples of special dictionary and list types that would potentially benefit from inferring and propagating the specific dictionary and list types are below.

```TS
// lists when using an overload that does not propagate the full list type
const stringListWithAddedBits = { 0: 'str', length: 1, nonListProperty: 'suffix' }; // compatible with List<string>
_(stringListWithAddedBits).value().nonListProperty; // error - nonListProperty does not exist
_(stringListWithAddedBits).each(value => console.log(value)).nonListProperty; // error - nonListProperty does not exist
_(stringListWithAddedBits).map((value, index, list) => value + list.nonListProperty); // error - nonListProperty does not exist

// dictionaries when using an overload that does not propagate the full dictionary type
const dictionaryWithKnownProperties = {
    numberProperty: 3,
    alsoANumberProperty: 4
}; // compatible with Dictionary<number>
_(dictionaryWithKnownProperties).value().numberProperty; // error - numberProperty does not exist
_(dictionaryWithKnownProperties).each(value => console.log(value)).numberProperty; // error - numberProperty does not exist
_(dictionaryWithKnownProperties).map((value, index, list) => value + list.numberProperty); // error - numberProperty does not exist
```

Reply:
Actually, I'm mildly convinced. I agree that `_(x).value()` is useless, but the other bits are nice to have. I'd approve keeping this change.

Reply:
Sounds good, I suppose it would be good to add some tests around these cases as well so I'll go do that 😄

Reply:
Surprisingly when putting together these tests I found that an automatically-typed new object can be inferred to be a dictionary with string properties but one with an interface containing only string properties cannot, which admittedly makes the dictionary case outlined above a little less useful.

```TS
// I tried using both non-string and string syntax when defining the property names below just to see if one would make TS happier than the other
interface StronglyKeyedSimpleStringObjectDictionary {
    'a': SimpleStringObject;
    'b': SimpleStringObject;
    'c': SimpleStringObject;
}

const rawSimpleSimpleObjectDictionary = { a: { a: 'a', b: 'c' }, b: { a: 'b', b: 'b' }, c: { a: 'c', b: 'a' } };
const stronglyKeyedSimpleStringObjectDictionary: StronglyKeyedSimpleStringObjectDictionary = rawSimpleSimpleObjectDictionary;
const simpleStringObjectDictionary1: _.Dictionary<SimpleStringObject> = rawSimpleSimpleObjectDictionary; // ok
const simpleStringObjectDictionary2: _.Dictionary<SimpleStringObject> = stronglyKeyedSimpleStringObjectDictionary; // error: Type 'StronglyKeyedSimpleStringObjectDictionary' is not assignable to type 'Dictionary<SimpleStringObject>'. Index signature is missing in type 'StronglyKeyedSimpleStringObjectDictionary'.
```

That being said, it's still possible to define an interface that has properties that are known to be defined as well as an index signature, which could still be useful, so I added tests around that to close this out.